src/server.py

```diff
@@ -0,0 +1,168 @@
+            if directory == path:
+                return os.path.join(root, directory)
+        for file in files:
+            if file == path:
+                return os.path.join(root, file)
+
+
+def resolve_uri(uri):
+    """Take a URI parsed from a request.
+
+    It will return a body for a response with the type of
+    content contained in the body.
+    """
+    path = get_path(uri.split('/')[-1]) if uri != '/' else walk_dir
+    if path and os.path.isfile(path):
+        with open(path, mode='rb') as file:
+            file_content = file.read()
+            mime = MimeTypes()
+            content_type = mime.guess_type(path)[0]
+            return file_content, str(content_type).encode()
+    elif path and os.path.isdir(path):
+        contents = os.listdir(path)
+        files = []
+        folders = []
+        folder_path = path.split('webroot')[-1]
+        for x in contents:
+            if os.path.isfile(os.path.abspath(x)):
+                files.append(x)
+            else:
+                folders.append(x)
+        return str(folder_contents_html(folder_path, files, folders)).encode(), b'text/html'
+    else:
+        raise ValueError(404)
+
+
+def handle_message(conn, buffer_length):
+    """Handle the messages coming into the server."""
+    conn.setblocking(1)
+    message = []
+    while True:
+        part = conn.recv(buffer_length)
+        message.append(part)
+        print('Receiving message from client...\n')
+        if len(part) < buffer_length or part[-2:] == b'\r\n':
+            break
+        else:
+            print('Hold on, there is more...Receiving...')
+    print('parsing request...\n')
+
+    full_message = b''.join(message)
+    try:
+        uri = parse_request(full_message.decode('utf8'))
+        body, content_type = resolve_uri(uri)
+        message = response_ok(body, content_type)
+    except ValueError as e:
+        message = response_error(*e.args).encode('utf8')
+    except IndexError:
+        message = response_error(400).encode('utf8')
+    print('Sending Response...\n')
+    conn.sendall(message)
+    conn.close()
+
+
+def initialize_connection():
+    """Set up a socket a connection and return socket object."""
+    server = socket.socket(socket.AF_INET,
+                           socket.SOCK_STREAM,
+                           socket.IPPROTO_TCP)
+    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    server.bind(address)
+    server.listen(1)
+    print('\n', 'Server running on http://127.0.0.1:{}/'.format(address[1]), '\n')
+    return server
+
+
+def server():
+    """Start the server to listen and accept http requests."""
+    server = initialize_connection()
+
+    while True:
+        try:
+            conn, addr = server.accept()
+            print('Received a connection from: ', addr)
+            handle_message(conn, buffer_length)
+        except KeyboardInterrupt:
+            print('\n\nShutting Down Server...')
+            server.close()
```

Comment:
You should actually shut the server down before closing it so that you can send a packet back to the client signaling that the server is done.